### ISSUE TITLE
`verifyPhoneNumber` errors with `auth/invalid-app-credential` only on localhost

### ISSUE THREAD

**Issue by MaciejCaputa** [firebase/firebase-js-sdk#8387]
### Operating System

macOS Sonoma 14.2.1

### Browser Version

Chrome 126.0.6478.183 

### Firebase SDK Version

10.12.4

### Firebase SDK Product:

AppCheck, Auth

### Describe your project's tooling

Next.js using firebase app check with reCAPTCHA Enterprise with configured debug app check token

### Describe the problem

On localhost `verifyPhoneNumber` calls following endpoint

`https://identitytoolkit.googleapis.com/v2/accounts/mfaSignIn:start?key=<api-key>`

which returns

```
{
  "error": {
    "code": 400,
    "message": "INVALID_APP_CREDENTIAL",
    "status": "INVALID_ARGUMENT"
  }
}
```

### Steps and code to reproduce issue

- Interestingly,`verifyPhoneNumber` work perfectly fine, but stopped working couple of days ago only when developing locally on localhost
- The deployed (hosted on Vercel) even after redeployment continue to work just fine
- Similar issue reported here https://www.reddit.com/r/Firebase/comments/1e8pdcf/firebase_error_authinvalidappcredentials_in/

**Comment by odina101** [comment]
firebase has issue on its servers

**Comment by shahsagarm** [comment]
I'm also facing the same exact issue. It suddenly stopped working locally but on deployed version (vercel) it works fine with same firebase config.

**Comment by MaciejCaputa** [comment]
> firebase has issue on its servers

Do they know they have an issue? Are they working on the fix? Is there a ticket we can track?

**Comment by Sam-Persoon** [comment]
We also have a ticket opened at GCP. For those that want a temporary workaround, you can serve on `127.0.0.1` instead of `localhost` and add `127.0.0.1` in Firebase console -> authentication -> settings -> authorised domains.

**Comment by LizDodion** [comment]
Also seeing this

I added 127.0.0.1 as an authorised domain and switched our app to do that and it worked fine

Would be great if firebase can fix asap as we have a lot of other apps that need to be reconfigured and don't want to have to add everywhere

**Comment by jbalidiong** [comment]
Hi @MaciejCaputa, thanks for reaching out to us. I was able to replicate the behavior. Let me check what we can do for this issue with our Auth team or bring someone here that can provide more context about it. I’ll update this thread if I have any information to share.

**Comment by deepakrh185** [comment]
I'm experiencing the same issue in my React project

**Comment by saharvx9** [comment]
Working only for test numbers, iam getting this error too in flutter with "signInWithPhoneNumber"

**Comment by Sumit-Mayani** [comment]
> I'm encountering the same issue in my React project. It seems to only occur in the web app, while it works fine on mobile.

its working in mobile correctly but issue face only on web

**Comment by Sumit-Mayani** [comment]
> We also have a ticket opened at GCP. For those that want a temporary workaround, you can serve on `127.0.0.1` instead of `localhost` and add `127.0.0.1` in Firebase console -> authentication -> settings -> authorised domains.

not working

**Comment by yeonjoonkim** [comment]
Hi I am having same problem with phone authentication on `localhost`.

### **I have included `localhost` and `127.0.0.1` into firebase console and google cloud**
1. Firebase Console => Authentication => Setting => Approved Domain 
- `localhost`
- `127.0.0.1`


2. Google Cloud => `Project` => Credential => OAuth 2.0 Client ID => Authroised Javascript ...
- `http://localhost`
- `https://localhost`
- `http://127.0.0.1`
- `https://127.0.0.1`
- `https://localhost:8100`
- `https://localhost:8101`
- `https://localhost:5000`


### **Error Code**

`https://identitytoolkit.googleapis.com/v1/accounts:sendVerificationCode?key=${key}`
`Request method: Post`
`Access-Control-Allow-Origin: http:localhost:8100`

```
Response: {
  "error": {
    "code": 400,
    "message": "INVALID_APP_CREDENTIAL",
    "errors": [
      {
        "message": "INVALID_APP_CREDENTIAL",
        "domain": "global",
        "reason": "invalid"
      }
    ]
  }
}
```


### **Environment**

`package`

```
    "@angular/animations": "^18.1.1",
    "@angular/cdk": "^18.1.1",
    "@angular/common": "^18.1.1",
    "@angular/core": "^18.1.1",
    "@angular/fire": "^18.0.1",
    "@angular/forms": "^18.1.1",
    "@angular/localize": "^18.1.1",
    "@angular/platform-browser": "^18.1.1",
    "@angular/platform-browser-dynamic": "^18.1.1",
    "@angular/router": "^18.1.1",
    "@angular/service-worker": "^18.1.1",
    "firebase": "10.12.4",
```

Phone Authentication Code

```
  public async sendOTP(ph: string): Promise<void> {
    try {
      this.registerRecaptchaDiv();
      const verificationId = await this.getVerificationId(ph);
      this.phoneCodeSentEvent$.next(verificationId);
    } catch (error) {
      this.errorMsg$.next(getErrorMsg(error));
    }
  }
  
    public async confirmOTP(verificationId: string, verificationCode: string): Promise<void> {
    try {
      const authCredential = PhoneAuthProvider.credential(verificationId, verificationCode);
      const userCredential = await signInWithCredential(this.auth, authCredential);

      this.signInResult$.next(userCredential?.user ?? null);
    } catch (error) {
      this.errorMsg$.next(getErrorMsg(error));
    }
  }
    
    private registerRecaptchaDiv() {
    const recaptchContainer = document.getElementById('recaptcha-container');

    if (recaptchContainer) {
      const parent = recaptchContainer.parentNode;
      parent?.removeChild(recaptchContainer);

      const newRecaptchContainer = document.createElement('div');
      newRecaptchContainer.id = 'recaptcha-container';
      parent?.appendChild(newRecaptchContainer);
    }
  }
  
  
    private getVerificationId(ph: string) {
    const appVerifier = new RecaptchaVerifier(this.auth, 'recaptcha-container', {
      size: 'invisible',
      callback: (response: any) => {
        console.log('Recaptcha verified');
      },
    });

    const provider = new PhoneAuthProvider(this.auth);
    const verificationId = provider.verifyPhoneNumber(ph, appVerifier);
    appVerifier.clear();
    return verificationId;
  }
  
```

**Comment by diffractometer** [comment]
+1 seeing this issue a lot lately

**Comment by DellaBitta** [comment]
Hi all,

This error is due a change in Auth Service policy which no longer supports localhost for `verifyPhoneNumber` invocations. The new guidance is part of [our documentation here](https://firebase.google.com/docs/auth/web/phone-auth#enable-phone-number-sign-in-for-your-firebase-project).

The Auth team has suggested the use of  [fictional phone numbers](https://firebase.google.com/docs/auth/web/phone-auth#test-with-fictional-phone-numbers) for testing purposes.

I'm going to remove the `bug` label from this issue but I'll leave it open for any questions that may arise. Thanks.

**Comment by ansh** [comment]
Can we get some context around why localhost is no longer allowed? @DellaBitta

**Comment by yeonjoonkim** [comment]
@DellaBitta 

Hi If we create a fictional phone auth, do they have prefixed id?

Or

Is there have any firebase sdk that allow us to create fictional phone auth by firebase function?

**Comment by diffractometer** [comment]
echoing @ansh kinda bogus but thanks for the update

**Comment by willemjanvankranenburg** [comment]
@DellaBitta Can you elaborate why this is changed? A lot (if not all) of hybrid apps use localhost as the url to show the pages  in the android/iOS web view, even when it's a signed production build. This is not something that a normal developer can change.

Also when having a user base that can use their own phone number to login simply saying "use of [fictional phone numbers](https://firebase.google.com/docs/auth/web/phone-auth#test-with-fictional-phone-numbers) for testing purposes" is **not** acceptable. 

It's also not acceptable that this is changed by the firebase auth team without notifying users. We have hybrid apps in production for hundreds of business users. They suddenly cannot login anymore and we need to hotfix a workaround for them.

**Comment by Sam-Persoon** [comment]
Agreeing with @willemjanvankranenburg, this indeed requires a sudden code change in a lot of moving parts in our ecosystem. Because no one was notified, there were also a lot of resources spend into researching the issue in the first place.

> The new guidance is part of [our documentation here](https://firebase.google.com/docs/auth/web/phone-auth#enable-phone-number-sign-in-for-your-firebase-project).

@DellaBitta it is quite ironic/confusing as `localhost` was (is?) added as an authorised domain by default.

**Comment by MJSPollard** [comment]
I am having this issue. Not even using the firebase SDK though. I am directly calling the identity toolkit google API's from my Node backend and locally it is failing every time now even though I have not changed anything in my app in weeks.

**Comment by Hilton-Borah** [comment]
Same for me also. I am facing this issue from last two days. Finally I get some response.

Can anyone explain about this, I am new in firebase 

Cross origin redirect sign-in on Google Chrome M115+ is no longer supported, and will stop working on June 24, 2024.
Migrate to an alternative solution before June 24, 2024
End-users can continue to sign in to your app until June 24, 2024
This is already required on Firefox 109+ and Safari 16.1+

Will it affect on auth services ?

**Comment by pnede3** [comment]
Tonight I wanted to update my nextjs app with 2fa by SMS, and after a few hours I found this. Can you please update the guide on 2fa by SMS with some information that localhost is not working?

**Comment by kamasuPaul** [comment]
Same issue here. Using Quasar framework with capacitor. It works on web(hosted version), but it doesn't work in the mobile app (that uses localhost)

**Comment by KhoubraneX** [comment]
I had the same issue while developing my app, and I believe the problem was related to using localhost. Here are the steps I followed to resolve it:

I used **Visual Studio Code's port forwarding** feature, which can be found in the **Remote Explorer panel**, to change from `localhost:port` to a URL, like `https://x1lpjwcv-port.euw.devtunnels........`

I added the domain provided by Visual Studio Code to Firebase by navigating to `Firebase Console, Authentication > Settings > Authorized Domains` and entering the domain there.

After doing this, everything worked fine.

**Comment by touy** [comment]
I have the same issues with localhost and ionic capacitor project, it was working before , 
now it has the issue : FirebaseError: Firebase: Error (auth/invalid-app-credential).
with testing phonenumber just work fine , 
not working with new phonenumber .

**Comment by jhonny-ventiades** [comment]
Any solution with this issue? I have an application that works with Ionic and Capacitor, we have already tried all the solutions proposed in the thread, but the problem persists. I currently have that authentication method not working and many of my users cannot login in the app.

**Comment by NhienLam** [comment]
Hi everyone,

We implemented a change to disallow Phone Auth from localhost to address a potential vulnerability that could have led to exploitation. While we recognize this change may have unintended consequences from some developers, we cannot rollback the fix due to the security risk it poses.

Impacted project owners will receive communication on next steps soon. We apologize for any inconvenience this has caused.

**Comment by aem** [comment]
@NhienLam how are users of the SDK expected to test implementations if not on localhost? Am I expected to write an authentication path and then ship it to prod without testing it? This is occurring even when using the configured test phone numbers.

**Comment by NhienLam** [comment]
Hi @aem, for local development testing, you should be able to use configured [fictional phone numbers](https://firebase.google.com/docs/auth/web/phone-auth#test-with-fictional-phone-numbers) or [Auth Emulator](https://firebase.google.com/docs/emulator-suite/connect_auth). Additionally, you can use `127.0.0.1` to test as mentioned above. 

>This is occurring even when using the configured test phone numbers.

You won't see `auth/invalid-app-credential` error when using the configured fictional phone number.

**Comment by aem** [comment]
I've been using fictional phone numbers and they do not work, I get the same error.

**Comment by NhienLam** [comment]
@aem that's odd. We verified that using fictional phone numbers worked. 

1. Could you try sending the request with "+15555555555" instead of "+1 555-555-5555"?
2. Could you make sure you configured the fictional phone number in the correct project that you were testing?

**Comment by touy** [comment]
> Hi everyone,
> 
> We implemented a change to disallow Phone Auth from localhost to address a potential vulnerability that could have led to exploitation. While we recognize this change may have unintended consequences from some developers, we cannot rollback the fix due to the security risk it poses.
> 
> Impacted project owners will receive communication on next steps soon. We apologize for any inconvenience this has caused.

oh, this is a bad dream all ionic project use a webview with localhost , so all ionic capacitor would be a trash , I have checked my website could work well with a proper domain , but it could not work for ionic capacitor project , that is a bad dream

**Comment by rscotten** [comment]
@NhienLam Why is it that `127.0.01` is allowed but `localhost` isn't?  Wouldn't they have the same security vulnerabilities?  

Also, whoever rolled out this update could have saved thousands of developers many man hours just by having a more informative error message like "localhost is no longer allowed for Multi Factor phone authentication.  Use 127.0.0.1 instead.  Read more <link>".

If Google/Firebase actually cares about their users, they should really consider the economic cost of bad error messages like this before rolling them out unannounced.

**Comment by kondricz** [comment]
Hello,

Anyone else having this issue coming up in a hosted environment?

I have not found a pattern yet, but the error is showing up in our systems even when deployed. I have not found a reproducable pattern yet, but some of our developers are facing this issue regularly, especially in in-app browsers (like browser from instagram, browser from telegram).

We use nextJS.

**Comment by MaciejCaputa** [comment]
> Hello,
> 
> Anyone else having this issue coming up in a hosted environment?
> 
> I have not found a pattern yet, but the error is showing up in our systems even when deployed. I have not found a reproducable pattern yet, but some of our developers are facing this issue regularly, especially in in-app browsers (like browser from instagram, browser from telegram).
> 
> We use nextJS.

We also use NextJS and experience that on products sometimes (15% of times) SMS codes are not delivered. We have not yet verified if it is because of this issue, or another issue, but I can't imagine mobile carriers loosing around 15% of SMS.

Can anybody advise how we can report it?

**Comment by antavi360** [comment]
It is beyond my belief that Google just changes these things unilaterally without proper communication / documentation.
Aside of the 'localhost' issue we are experiencing from a good number of users with that same error on production. 
Suddently without touching the code.

The only solution for us now is to throw out this gcp auth garbage and do our own flows.
Google developers need to understand that such breaking changes or whatever this is are hurting mission critical projects. 
And there is no room for slack for us.

**Comment by gvillaume** [comment]
For anyone using ionic capacitor, the localhost issue can be fixed in production by updating the server hostname in your `capacitor.config.ts` file.

Ex.
```ts
const config: CapacitorConfig = {
...
  server: {
    hostname: 'my-app.com',
    androidScheme: 'https',
  }
};
```

Here's the documentation: https://capacitorjs.com/docs/guides/autofill-credentials#set-capacitor-server-hostname

**Comment by diffractometer** [comment]
+1 seeing this in production a fair amount `auth/invalid-app-credential` cc @kondricz

**Comment by kondricz** [comment]
Hey everyone,

Did Google rolled back this update? I just tested it now (with non-testing real number) on localhost, and it is working again finally.

**Comment by LizDodion** [comment]
It started working back again for me today

**Comment by antavi360** [comment]
> It started working back again for me today after not working for a few weeks. hopefully it is now rolled back

I can't confirm this. After performing the recaptcha, the error is still appearing for localhost and unclear if it still appears in production. Waiting on client feedback.

ERROR 1: identitytoolkit.googleapis.com/v1/accounts:sendVerificationCode?key=<KEY>:1 httperror 400 Failed to load resource: the requested URL /v1/accounts:sendVerificationCode?key=<KEY> was not found on this server. 
ERROR 2:  recaptcha__en.js:249 Uncaught TypeError: Cannot read properties of null (reading 'style')

firebase_core: ^2.25.4
firebase_auth: ^4.17.4

**Comment by kondricz** [comment]
I think they are working on it right now. Can someone confirm this from google team?

I had 5 sign ups and it was also working from localhost, but now the error is appearing again. This is VERY BAD. Why can't we just roll it back and release it only when its properly tested and won't break all of our stuff ?

**Comment by mohsenkhkh** [comment]
> > We also have a ticket opened at GCP. For those that want a temporary workaround, you can serve on `127.0.0.1` instead of `localhost` and add `127.0.0.1` in Firebase console -> authentication -> settings -> authorised domains.
> 
> I tried `127.0.0.1` and added it as an authorized domain, but it didn't work for me. Perhaps it is because we use reCaptcha Enterprise where we don't have whitelisted `127.0.0.1`

did you place 127.0.0.1 instead of localhost in your browser too? that was for me 127.0.0.1:3000 instead of localhost:3000

**Comment by kondricz** [comment]
Any updates here? Our team is still suffering from this.

We are considering just to build this thing on our own or use another provider.

**Comment by trauma-and-drama** [comment]
There seems to be an issue not only with localhost, but potentially proxy or NAT-based networks (can't tell).

We pulled the plug and deployed a custom Phone-Auth via Twillio, Firebase https-Functions, and customTokens.
This whole handshake with recaptcha is just too buggy and blackboxy for us, the settings page for recaptcha does not seem to be well integrated. And the support here, is mindboggingly absent on a critical issue like this (3 weeks!!)

My advice to the Google Firebase/GCP team:
If you want small businesses to use your services, take more care on breaking changes. 
The world is more complex beyond your Google cubicle (this is me being nice and mindful).

**Comment by pth55** [comment]
> Hi I am having same problem with phone authentication on `localhost`.
> 
> ### **I have included `localhost` and `127.0.0.1` into firebase console and google cloud**
> 1. Firebase Console => Authentication => Setting => Approved Domain
> 
> * `localhost`
> * `127.0.0.1`
> 
> 2. Google Cloud => `Project` => Credential => OAuth 2.0 Client ID => Authroised Javascript ...
> 
> * `http://localhost`
> * `https://localhost`
> * `http://127.0.0.1`
> * `https://127.0.0.1`
> * `https://localhost:8100`
> * `https://localhost:8101`
> * `https://localhost:5000`
> 
> ### **Error Code**
> `https://identitytoolkit.googleapis.com/v1/accounts:sendVerificationCode?key=${key}` `Request method: Post` `Access-Control-Allow-Origin: http:localhost:8100`
> 
> ```
> Response: {
>   "error": {
>     "code": 400,
>     "message": "INVALID_APP_CREDENTIAL",
>     "errors": [
>       {
>         "message": "INVALID_APP_CREDENTIAL",
>         "domain": "global",
>         "reason": "invalid"
>       }
>     ]
>   }
> }
> ```
> 
> ### **Environment**
> `package`
> 
> ```
>     "@angular/animations": "^18.1.1",
>     "@angular/cdk": "^18.1.1",
>     "@angular/common": "^18.1.1",
>     "@angular/core": "^18.1.1",
>     "@angular/fire": "^18.0.1",
>     "@angular/forms": "^18.1.1",
>     "@angular/localize": "^18.1.1",
>     "@angular/platform-browser": "^18.1.1",
>     "@angular/platform-browser-dynamic": "^18.1.1",
>     "@angular/router": "^18.1.1",
>     "@angular/service-worker": "^18.1.1",
>     "firebase": "10.12.4",
> ```
> 
> Phone Authentication Code
> 
> ```
>   public async sendOTP(ph: string): Promise<void> {
>     try {
>       this.registerRecaptchaDiv();
>       const verificationId = await this.getVerificationId(ph);
>       this.phoneCodeSentEvent$.next(verificationId);
>     } catch (error) {
>       this.errorMsg$.next(getErrorMsg(error));
>     }
>   }
>   
>     public async confirmOTP(verificationId: string, verificationCode: string): Promise<void> {
>     try {
>       const authCredential = PhoneAuthProvider.credential(verificationId, verificationCode);
>       const userCredential = await signInWithCredential(this.auth, authCredential);
> 
>       this.signInResult$.next(userCredential?.user ?? null);
>     } catch (error) {
>       this.errorMsg$.next(getErrorMsg(error));
>     }
>   }
>     
>     private registerRecaptchaDiv() {
>     const recaptchContainer = document.getElementById('recaptcha-container');
> 
>     if (recaptchContainer) {
>       const parent = recaptchContainer.parentNode;
>       parent?.removeChild(recaptchContainer);
> 
>       const newRecaptchContainer = document.createElement('div');
>       newRecaptchContainer.id = 'recaptcha-container';
>       parent?.appendChild(newRecaptchContainer);
>     }
>   }
>   
>   
>     private getVerificationId(ph: string) {
>     const appVerifier = new RecaptchaVerifier(this.auth, 'recaptcha-container', {
>       size: 'invisible',
>       callback: (response: any) => {
>         console.log('Recaptcha verified');
>       },
>     });
> 
>     const provider = new PhoneAuthProvider(this.auth);
>     const verificationId = provider.verifyPhoneNumber(ph, appVerifier);
>     appVerifier.clear();
>     return verificationId;
>   }
>   
> ```



> Hi I am having same problem with phone authentication on `localhost`.
> 
> ### **I have included `localhost` and `127.0.0.1` into firebase console and google cloud**
> 1. Firebase Console => Authentication => Setting => Approved Domain
> 
> * `localhost`
> * `127.0.0.1`
> 
> 2. Google Cloud => `Project` => Credential => OAuth 2.0 Client ID => Authroised Javascript ...
> 
> * `http://localhost`
> * `https://localhost`
> * `http://127.0.0.1`
> * `https://127.0.0.1`
> * `https://localhost:8100`
> * `https://localhost:8101`
> * `https://localhost:5000`
> 
> ### **Error Code**
> `https://identitytoolkit.googleapis.com/v1/accounts:sendVerificationCode?key=${key}` `Request method: Post` `Access-Control-Allow-Origin: http:localhost:8100`
> 
> ```
> Response: {
>   "error": {
>     "code": 400,
>     "message": "INVALID_APP_CREDENTIAL",
>     "errors": [
>       {
>         "message": "INVALID_APP_CREDENTIAL",
>         "domain": "global",
>         "reason": "invalid"
>       }
>     ]
>   }
> }
> ```
> 
> ### **Environment**
> `package`
> 
> ```
>     "@angular/animations": "^18.1.1",
>     "@angular/cdk": "^18.1.1",
>     "@angular/common": "^18.1.1",
>     "@angular/core": "^18.1.1",
>     "@angular/fire": "^18.0.1",
>     "@angular/forms": "^18.1.1",
>     "@angular/localize": "^18.1.1",
>     "@angular/platform-browser": "^18.1.1",
>     "@angular/platform-browser-dynamic": "^18.1.1",
>     "@angular/router": "^18.1.1",
>     "@angular/service-worker": "^18.1.1",
>     "firebase": "10.12.4",
> ```
> 
> Phone Authentication Code
> 
> ```
>   public async sendOTP(ph: string): Promise<void> {
>     try {
>       this.registerRecaptchaDiv();
>       const verificationId = await this.getVerificationId(ph);
>       this.phoneCodeSentEvent$.next(verificationId);
>     } catch (error) {
>       this.errorMsg$.next(getErrorMsg(error));
>     }
>   }
>   
>     public async confirmOTP(verificationId: string, verificationCode: string): Promise<void> {
>     try {
>       const authCredential = PhoneAuthProvider.credential(verificationId, verificationCode);
>       const userCredential = await signInWithCredential(this.auth, authCredential);
> 
>       this.signInResult$.next(userCredential?.user ?? null);
>     } catch (error) {
>       this.errorMsg$.next(getErrorMsg(error));
>     }
>   }
>     
>     private registerRecaptchaDiv() {
>     const recaptchContainer = document.getElementById('recaptcha-container');
> 
>     if (recaptchContainer) {
>       const parent = recaptchContainer.parentNode;
>       parent?.removeChild(recaptchContainer);
> 
>       const newRecaptchContainer = document.createElement('div');
>       newRecaptchContainer.id = 'recaptcha-container';
>       parent?.appendChild(newRecaptchContainer);
>     }
>   }
>   
>   
>     private getVerificationId(ph: string) {
>     const appVerifier = new RecaptchaVerifier(this.auth, 'recaptcha-container', {
>       size: 'invisible',
>       callback: (response: any) => {
>         console.log('Recaptcha verified');
>       },
>     });
> 
>     const provider = new PhoneAuthProvider(this.auth);
>     const verificationId = provider.verifyPhoneNumber(ph, appVerifier);
>     appVerifier.clear();
>     return verificationId;
>   }
>   
> ```

this method is working!! i have added 127.0.0.1 to the authorized domains list and replaced the same in local url

**Comment by kondricz** [comment]
It is funny for me.

It feels like the team is constantly working on the issue, but they sort of pushing on prod and let us end users test it 😭 

Sometimes it works flawlessly for hours or even half a day (I guess this is the time when they restore the older version of the API), then it randomly starts failing for few hours.

Yesterday we had around 100 phone registration and 5 of them failed with this. Is there anyone here who has not migrated to their own solution yet and experiencing the same?

**Comment by kyleabens** [comment]
When did this change from localhost come into effect? Is there a version of firebase we can roll back to?

**Comment by touy** [comment]
I can solve the problem by doing these 
1. server-side  Nodejs
    using version: "firebase-admin": "^12.2.0",
    
2. Client-side ionic capacitor:
https://github.com/firebase/firebase-js-sdk/issues/8387#issuecomment-2284610382

3. and client code ionic capacitor:
using version:    "firebase": "^9.23.0",

and you can keep your old codes  and it could work normally 

client-side firebase from npm changed completely and it needs to change a lot of works  , this is the shortest way I have found

**Comment by kyleabens** [comment]
> I can solve the problem by doing these
> 
> 1. server-side  Nodejs
>    using version: "firebase-admin": "^12.2.0",
> 2. Client-side ionic capacitor:
>    [`verifyPhoneNumber` errors with `auth/invalid-app-credential` only on localhost #8387 (comment)](https://github.com/firebase/firebase-js-sdk/issues/8387#issuecomment-2284610382)
> 3. and client code ionic capacitor:
>    using version:    "firebase": "^9.23.0",
> 
> and you can keep your old codes and it could work normally
> 
> client-side firebase from npm changed completely and it needs to change a lot of works , this is the shortest way I have found

This does work but changing the hostname in capacitor.config can lead to a lot of other issues for Android, just FYI.  I changed the hostname to 127.0.0.1 for the time being for 2FA to work again

**Comment by AkshayThota7747** [comment]
For local testing, use 127.0.0.1 and allow the same in authorized domain. It worked for me and it's always for me in the production. I was facing the issue only in the local environment

**Comment by rajeshpal53** [comment]
IP> @NhienLam Why is it that `127.0.01` is allowed but `localhost` isn't? Wouldn't they have the same security vulnerabilities?
> 
> Also, whoever rolled out this update could have saved thousands of developers many hours just by having a more informative error message like `localhost is no longer allowed for Multi Factor phone authentication. Use 127.0.0.1 instead. Read more (link)`.
> 
> If Google/Firebase actually cares about their users, they should really consider the economic cost of bad error messages like this before rolling them out unannounced.

You are right , its working with other ips but not by localhost, i wasted 1 week to find solutions and enable every possible solution to fix this, and its currently working with just change by localhost to any IP

**Comment by rajeshpal53** [comment]
> Any updates here? Our team is still suffering from this.
> 
> We are considering just to build this thing on our own or use another provider.

use any IP instead of localhost

**Comment by Gabe1203** [comment]
Has anyone been able to fix this issue on a hosted site? Our local works just fine withe a test number but in our hosted site in production we can no longer get into the site which is very alarming. We haven't made any updates to the login flow on our site.

**Comment by NMann** [comment]
Is there any update on this issue, as I am still facing this issue

**Comment by DiyazY** [comment]
I have an electron application using firebase. Auth logic works with the test number but throws `auth/invalid-app-credential` when I test it with the real number. I had also tried all possible ways to tweak it on a firebase console, however, it didn't work.

**Comment by narenkram** [comment]
same on my Vue app, so many hours wasted.🥲

**Comment by marianocodes** [comment]
> same on my Vue app, so many hours wasted.🥲

Same here

**Comment by timothyjoelwright** [comment]
I'm getting this error in production using public domains but only with certain live phone numbers. Test numbers and test devices with live numbers are working with the exception of a few specific real numbers.

**Comment by Prathiv02** [comment]
> I've also been a recent victim of this issue, probably wasted about 10 hours of my life. If only AI could solve issues like these 😅
> 
> Step 1: use `127.0.0.1:3000` instead of `locahost:3000`. Step 2: authorize the domain `127.0.0.1` on your Firebase console.

Can anyone confirm whether this workaround is working because , adding 127.0.0.1 and running app on 127.0.0.1 & added test phone number with hardcoded OTP also not working

**Comment by vishwa1004** [comment]
> I have the same issues with localhost and ionic capacitor project, it was working before , now it has the issue : FirebaseError: Firebase: Error (auth/invalid-app-credential). with testing phonenumber just work fine , not working with new phonenumber .

Hey you got any solution for this issue? actually i am also facing the same issue

**Comment by vishwa1004** [comment]
> We are also having this issue, with a Ionic Cordova app, since past friday. We have a P1 ticket opened at GCP support currently.

Hi, any solution for this?

**Comment by MiklerGM** [comment]
> Can anyone confirm whether this workaround is working because , adding 127.0.0.1 and running app on 127.0.0.1 & added test phone number with hardcoded OTP also not working

I've tried the following: 
1. Added a new custom autorized domain for example `my.app` (you can write here anything you want)
2. Added a new value to `/etc/hosts` on my local machine e.g. `127.0.0.1 my.app`
3. Opened the application in browser using `my.app:<port>` instead of `localhost:<port>`
4. Enjoy working sms

**Comment by muthu32** [comment]
Sms sending works on test numbers but not works on Real numbers, Because of Domain Validation issue.
Why Can't firebase throw domain errors on test number authentication? It will show error on production and it will not show error on test numbers. Then what is the use of test numbers here? Are those test numbers are joke to those firbase team???
It is clearly confusing and makes every new developers waste their time on those issues
Here my juniors are wondering why test number is working but not real numbers, and error here is thrown was even bigger joke, telling INVALID_APP_CREDENTIAL, instead of invalid domain, ha ha..

**Comment by vivek1023** [comment]
> I've also been a recent victim of this issue, probably wasted about 10 hours of my life. If only AI could solve issues like these 😅
> 
> Step 1: use `127.0.0.1:3000` instead of `locahost:3000`. Step 2: authorize the domain `127.0.0.1` on your Firebase console.

you save me, bro thanks

**Comment by diffractometer** [comment]
my recommendation is to stop using firebase 😓

**Comment by viveklele** [comment]
Do you know any alternative?

**Comment by gyanraj06** [comment]
Same issue here - still. This time on prod as well as on local. Entering the IP in authorized domain also did not help.

**Comment by viveklele** [comment]
Instead of localhost try to use ip
127.0.0.1:port_number

**Comment by lex-gsynergy** [comment]
When authorizing 127.0.0.1, I found that the Firebase Console will display an error message about it not being a valid host name. However, you can still click the Add button despite the error message.

Like many, I found that switching to 127.0.0.1 fixed the problem. It seems like a bug in Firebase, but this is a good workaround. It may be worth at documenting the workaround in the main documentation for Firebase, unless or until it is possible to use "localhost".

**Comment by udaysamandsar** [comment]
Reported by: Uday Bhan
Email: udaysamandsar@gmail.com

Issue

Firebase Phone Authentication was failing during local development with:

auth/invalid-app-credential

and later:

auth/captcha-check-failed
Hostname match not found
Root Cause

The app was running on:

http://127.0.0.1:3000

but 127.0.0.1 was not added in:

Firebase Console
→ Authentication
→ Settings
→ Authorized Domains
Solution

Add the following domain:

127.0.0.1

(Do not add 127.0.0.1:3000)

Run the app with:

npm run dev -- --host 127.0.0.1

and open:

http://127.0.0.1:3000
Result
✅ reCAPTCHA verified successfully.
✅ signInWithPhoneNumber() worked.
✅ OTP was sent successfully.
✅ Phone Authentication worked as expected.

Note: This only affects local development and does not impact production websites or deployed applications.
